kotlin-coroutines-end/app/src/test/java/com/example/android/kotlincoroutines/main/MainViewModelUnitTest.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.android.kotlincoroutines.main
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import com.example.android.kotlincoroutines.CoroutinesMainDispatcherRule
+import com.example.android.kotlincoroutines.LiveDataTestUtil
+import com.example.android.kotlincoroutines.fakes.MainNetworkFake
+import com.example.android.kotlincoroutines.fakes.TitleDaoFake
+import com.example.android.kotlincoroutines.util.FakeNetworkCall
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Rule
+import org.junit.Test
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+
+class MainViewModelUnitTest {
+
+    // Subject under test
+    private lateinit var mainViewModel: MainViewModel
+
+    // A CoroutineDispatcher that can be controlled from tests
+    @ExperimentalCoroutinesApi
+    private val testDispatcher = TestCoroutineDispatcher()
+
+    // Set the main coroutines dispatcher for unit testing.
+    // We are setting the above-defined testDispatcher as the Main thread dispatcher.
+    @ExperimentalCoroutinesApi
+    @get:Rule
+    var coroutinesMainDispatcherRule = CoroutinesMainDispatcherRule(testDispatcher)
+
+    // Executes each task synchronously using Architecture Components.
+    @get:Rule
+    val instantTaskExecutorRule = InstantTaskExecutorRule()
+
+    @Test
+    fun refreshTitle() {
+        mainViewModel = MainViewModel(
+                TitleRepository(
+                        MainNetworkFake(FakeNetworkCall()),
+                        TitleDaoFake("title")
+                )
+        )
+        mainViewModel.refreshTitle()
+
+        assertThat(LiveDataTestUtil.getValue(mainViewModel.title)).isEqualTo("title")
```

Comment:
This can be done immediately if you use runBlockingTest or explicitly control time on the dispatcher while there is an active test observer. This is both more straightforward and deterministic.